### PR TITLE
No more weird caching issues.

### PR DESCRIPTION
**Pull request by brycedrennan**
Our pull request to the parent library has been ignored. Lets deploy to our private repo.

https://github.com/john-kurkowski/tldextract/pull/144

This is already deployed to pypi.cu and in use now.
http://pypi.cu/root/circleup/tldextract/3.0.0.circleup

### CODE REVIEW

**Review by jessevogt — CHANGES_REQUESTED** [review]
I am not comfortable with forking at this time since we would either lose access to new fixes from upstream OR would need to take on the work of manually keeping in sync. I would rather investigate implementing these fixes on our side external to tldextract.

**Comment by brycedrennan** [comment]
I do wish I had explored that route more as that would have been easier to maintain. There is some reason to think its not possible though because of how the library is designed.

I still think we should fork, as we have done in other cases because:
 - This solves a problem that has come up repeatedly for multiple people on our team and is very confusing when it happens.  The problem is that whether you choose to include private domains is cached "forever".  If the next call to the library specifies something different than the first call, then it's silently ignored, and incorrect results are returned.   This is totally unexpected and hard to debug.  It also  totally precludes different parts of code using different lists for different purposes.
- I think its unlikely we'll need the upstream changes.  The library is an overcomplicated way of downloading a text file of regular expressions.  It's already doing what we need.
- It's not clear there will even be upstream changes. There have been no commits for 9 months. No response to this PR. 
- These changes have been live for a month.

File: tldextract/utils.py [hunk 1]
```diff
@@ -22,12 +23,16 @@ def return_cache(*args, **kwargs):
...
+
+            # without locking, Concurrency could lead to multiple writers, or readers of a partial write
+            with FileLock(lock_path, timeout=20):
+                if not os.path.isfile(cache_path):
+                    result = func(*args, **kwargs)
+                    with open(cache_path, 'w') as cache_file:
+                        json.dump(result, cache_file)
 
             with open(cache_path) as cache_file:
                 return json.load(cache_file)
```
Comment (jessevogt) [inline tldextract/utils.py]: shouldn't the read also be protected by the lock?

**Review by jessevogt — CHANGES_REQUESTED** [review]
If backing out these changes and fixing via the method I mentioned in my initial comment is not an option, please update readme for this repo to mention how/when this fork happened + additional relevant background around the changes we introduced. We should either remove or update the badges as well since they still point to the upstream version.

File: tldextract/utils.py (hunk 1 above)
Comment (brycedrennan) [reply tldextract/utils.py]: good catch will fix

**Comment by jessevogt** [comment]
Pleas also update permissions on this repo. I am currently seeing the following message:

image

**Comment by brycedrennan** [comment]
good feedback. will do.

**Comment by brycedrennan** [comment]
@jessevogt It locks for reads now and the team has access to the repo.

**Review by jessevogt — APPROVED** [review]

**Pull request merged**